firmware/lib/Gantry/Movement.cpp

```diff
@@ -1,54 +1,64 @@
-        axis->vel_profile_cur_trap[0] = counts_accel;
-        axis->vel_profile_cur_trap[1] = counts_hold;
-        axis->vel_profile_cur_trap[2] = counts_decel;
-    }
-    else if (dir == DIR_NEGATIVE) {
-        axis->vel_profile_cur_trap[0] = - counts_accel;
-        axis->vel_profile_cur_trap[1] = - counts_hold;
-        axis->vel_profile_cur_trap[2] = - counts_decel;
-    }
-
-    axis->encoder.desired = axis->encoder.current + axis->vel_profile_cur_trap[0];
-
-    DEBUG_PRINT_VAL("Desired count", axis->encoder.desired);
-    DEBUG_PRINT_VAL("Current count", axis->encoder.current);
-    start_axis(axis);
+    // Convert from steps to counts
+    uint32_t accel = shape.accel * units.counts_per_rev / units.steps_per_rev;
+    uint32_t vel_start = shape.vel_start * units.counts_per_rev / units.steps_per_rev;
+    uint32_t vel_hold = shape.vel_hold * units.counts_per_rev / units.steps_per_rev;
+
+    // Generate a velocity profile for the correct axis
+    VelProfile profile;
+    bool valid_profile = generate_vel_profile(accel, vel_start, vel_hold, total_counts, &profile);
+    if (!valid_profile) return false;
+
+    AxisMotion motion = {
+        .dir = dir,
+        .vel_start = shape.vel_start,
+        .vel_hold = shape.vel_hold,
+        .accel = shape.accel,
+        .counts_accel = (dir == DIR_POSITIVE ? (int32_t)profile.dist_accel : -1*(int32_t)profile.dist_accel),
+        .counts_hold  = (dir == DIR_POSITIVE ? (int32_t)profile.dist_hold  : -1*(int32_t)profile.dist_hold),
+        .counts_decel = (dir == DIR_POSITIVE ? (int32_t)profile.dist_decel : -1*(int32_t)profile.dist_decel),
+    };
+
+    // TODO actually return the AxisResult here
+    return (axis_start(axis_id, &motion) == AXIS_OK);
 }
 
-void home_axis(Axis *axis)
+bool move_axis_home_a(AxisId axis_id, MotionShape shape, MotionUnits units)
 {
-    axis->homing = true;
-    // math to calculate number of accel counts
-    uint32_t counts_accel = calc_counts_accel(axis->accel, axis->vel_min, VELOCITY_HOMING);
-    // check if the limit switched is pressed at home
-    LimitSwitchStatus status = (LimitSwitchStatus)digitalRead(axis->ls_home.pin);
-    DEBUG_PRINT_VAL("LS STATUS", status);
-    if (status == PRESSED) {
-        DEBUG_PRINT_VAL("DRIVING IN POSITIVE", 0);
-        axis->encoder.current = 0;
-        axis_trapezoidal_move_rel(axis, counts_accel, UINT32_MAX/2, counts_accel, DIR_POSITIVE); // move until limit switch is released
-        return;
-    }
-    else {
-        DEBUG_PRINT_VAL("DRIVING IN NEGATIVE", 0);
-        axis->encoder.current = UINT32_MAX;
-        axis_trapezoidal_move_rel(axis, counts_accel, UINT32_MAX/2, counts_accel, DIR_NEGATIVE);
-        return;
-    }
-}
\ No newline at end of file
+    return move_axis_rel(axis_id, DIR_NEGATIVE, INT32_MAX, shape, units);
+}
+
+bool move_axis_home_b(AxisId axis_id, MotionShape shape, MotionUnits units)
+{
+    return move_axis_rel(axis_id, DIR_POSITIVE, INT32_MAX, shape, units);
```

Comment:
This is always specifying a max possible number of count to move, until the limit switch is reached?

Reply:
Yeah, the idea is try to move forever, the limit switch will automatically stop you once you've reached the home.